crates/runtime/src/memory.rs

```diff
@@ -137,22 +137,25 @@ impl RuntimeLinearMemory for MmapMemory {
         self.maximum
     }
 
-    fn grow_to(&mut self, new_size: usize) -> Option<()> {
+    fn grow_to(&mut self, new_size: usize) -> Result<()> {
+        if new_size == usize::MAX {
+            bail!("impossible memory size")
+        }
+
         if new_size > self.mmap.len() - self.offset_guard_size - self.pre_guard_size {
             // If the new size of this heap exceeds the current size of the
             // allocation we have, then this must be a dynamic heap. Use
             // `new_size` to calculate a new size of an allocation, allocate it,
             // and then copy over the memory from before.
             let request_bytes = self
                 .pre_guard_size
-                .checked_add(new_size)?
-                .checked_add(self.extra_to_reserve_on_growth)?
-                .checked_add(self.offset_guard_size)?;
+                .checked_add(new_size)
+                .and_then(|s| s.checked_add(self.extra_to_reserve_on_growth))
+                .and_then(|s| s.checked_add(self.offset_guard_size))
+                .ok_or_else(|| format_err!("overflow"))?;
```

Comment:
Mind expanding this a bit to explain that it's an overflow trying to calculate the size of the allocation of the linear memory?